compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -635,7 +635,26 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 self.codegen_consume(bx, place.as_ref())
             }
 
-            mir::Operand::Constant(ref constant) => self.eval_mir_constant_to_operand(bx, constant),
+            mir::Operand::Constant(ref constant) => {
+                let constant_ty = constant.ty();
+                // Make sure this type is actually passed as an immediate vector.
+                // (In particular, this excludes packed SIMD vectors.)
+                if constant_ty.is_simd() {
+                    let layout = bx.layout_of(constant_ty);
+                    return match layout.abi {
+                        Abi::Vector { .. } => {
+                            let (llval, ty) = self.immediate_const_vector(bx, constant);
+                            OperandRef {
+                                val: OperandValue::Immediate(llval),
+                                layout: bx.layout_of(ty),
+                            }
+                        }
+                        _ => self.eval_mir_constant_to_operand(bx, constant),
+                    };
+                }
+
+                self.eval_mir_constant_to_operand(bx, constant)
```

Comment:
You still have two call sites for `eval_mir_constant_to_operand` here.

If you early-return the `OperandRef` above, you can just have a single call site after the `if`.